Add background HPOS cleanup process

jorgeatorres opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
This PR adds a new tool to to WC > Status > Tools, that can be used to remove legacy data from orders. It triggers a background process (similar to the sync one) that removes data from the postmeta table for all orders. The new functionality is only available when HPOS is active and sync is disabled.

Closes #41915 & #41916.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build both the legacy assets and WC to make sure the JS and CSS changes are correct, by running `pnpm --filter='@woocommerce/plugin-woocommerce' build`.
1. Create a few orders in WC > Orders or use a tool such as Smooth Generator.
1. If necessary, run `wp wc cot sync` to sync orders.
1. Go to WC > Settings > Advanced > Features and make sure the HPOS datastore is selected and sync is disabled.
1. Go to WC > Orders and take note of a few order IDs.
1. Confirm the number of legacy order records on your site by running `wp post list --post_type=shop_order --format=count`. Take note of this number.
1. Confirm that your legacy records contain metadata by running `wp post meta list <order_id>` for a few of the order IDs from step 6.
1. Go to WC > Status > Tools and locate the "Clean up order data from legacy tables" tool.
   <img width="1353" alt="Screenshot 2024-01-26 at 17 23 12" src="https://github.com/woocommerce/woocommerce/assets/184724/9e8aeda5-22b8-4573-8706-0f2cca9ce7ca">
1. Click "Clear data" to start the process. The tool will now change to the following and kickstart a background process:
   <img width="1293" alt="Screenshot 2024-01-26 at 18 23 55" src="https://github.com/woocommerce/woocommerce/assets/184724/c4d8fa43-9d2b-4a19-b00f-586da2463ad8">
1. Feel free to test the "Cancel" functionality and restart the process as needed.
1. While the background process is running, you can run `wp post list --post_type=shop_order --format=count` to confirm that the number of orders from step 6 is decreasing. This should eventually reach 0.
1. Run `wp post list --post_type=shop_order_placehold --format=count` to confirm that all of the legacy records have been converted to placeholders.
1. Confirm that your legacy records now contain no metadata by running `wp post meta list <order_id>` with some of the order IDs from step 5.
1. Visit WC > Status > Tools again and confirm that the tool is disabled and that the button previously used to start the process now has a "No orders in need of cleanup" message:
   <img width="1342" alt="Screenshot 2024-01-26 at 17 30 11" src="https://github.com/woocommerce/woocommerce/assets/184724/15781576-384d-45cd-bb89-1b37d9bc442a">


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

jorgeatorres committed:
Make legacy data cleanup more performant

jorgeatorres committed:
Add legacy data cleanup processor

jorgeatorres committed:
Make legacy cleanup available in the UI

jorgeatorres committed:
Add some validation around data synchronizer / data cleanup settings

jorgeatorres committed:
Add some error handling

jorgeatorres committed:
Add changelog

jorgeatorres committed:
Make CI happy

jorgeatorres committed:
Improve option handling

jorgeatorres committed:
Add unit tests

jorgeatorres committed:
Partially revert “real time” cleanup

jorgeatorres committed:
Re-schedule cleanup on ‘shutdown’ if necessary

vedanshujain commented:
@jorgeatorres , wdyt about making this "more" hidden, perhaps inside the Tools sections (but add a link to that tool page here, or keep it CLI only)? Otherwise, we may display a prompt when it has been some time (or some orders since HPOS was enabled). This is because:
1. I don't know if merchants who are more focussed on business would know the implications of this tool. (which is only that revert will take more time).
2. This seems complex, we should allow the merchant to track the progress, cancel it etc, basically anything that you can do from CLI. 

Perhaps @jarekmorawski would have more thoughts on how the UX should look like?

jorgeatorres commented:
Thanks for taking a look @vedanshujain!

> wdyt about making this "more" hidden, perhaps inside the Tools sections (but add a link to that tool page here, or keep it CLI only)?

As mentioned above, I'm open to suggestions on how to properly indicate the "seriousness" of this setting, but based on feedback we've received, I don't think it should be _too_ hard to enable or CLI only (though the CLI tool went out with 8.5.0). More than a few folks are waiting for a way to remove legacy data from their tables (for some, this was the main reason to switch), and even if the CLI tool does the job, there's probably value in making this available in the UI in a way that's easy to activate (provided they understand the consequences).

> This seems complex, we should allow the merchant to track the progress, cancel it etc, basically anything that you can do from CLI.

That makes sense. I can add some of this info to the settings screen, similarly to what we do with the data sync process.

Konamiman commented:
I also agree that a tool is more suited than a setting for a one-time operation like this one. If it's a background process you could even show progress in the tools page itself, like the "Regenerate the product attributes lookup table" does.

Maybe what could be added in the settings page, for the sake of discoverability, is a link to the tool.

jarekmorawski commented:
I agree with everything that's been said already. The clean-up functionality should be a tool with its own controls (start, stop, pause). We'd add a link to the settings page. Here's a quick mockup:

<img width="808" alt="image" src="https://github.com/woocommerce/woocommerce/assets/79307566/b1fcfcff-745e-4533-bf65-87b0198ff6f5">

<img width="879" alt="image" src="https://github.com/woocommerce/woocommerce/assets/79307566/532be996-7345-4032-b019-444f3603dbf2">

We should also show a success notice at the top of the screen once the clean-up is complete.

jorgeatorres commented:
@jarekmorawski: That looks awesome. Thanks for the mockups!

I'll work on moving this to the Tools section as discussed. Thank you @vedanshujain and @Konamiman for the suggestions.

jorgeatorres committed:
Revert CSS changes for HPOS settings screen

jorgeatorres committed:
Revert JS changes for HPOS settings screen

jorgeatorres committed:
Allow tools to display a status text

jorgeatorres committed:
Move order cleanup to Tools

jorgeatorres committed:
PHPCS fixes

jorgeatorres commented:
Hey @vedanshujain @Konamiman!

I think this is ready for another look. I've updated the testing instructions, but basically I've removed the new settings and moved the tool to WC > Status > Tools along with the other HPOS tools.

Remember to re-build the project so that you get the CSS/JS changes.

Here's a short screencast of how things are working now:

![Untitled](https://github.com/woocommerce/woocommerce/assets/184724/3ca0ab8b-d7d6-4d00-8fe7-4cd531a024b2)

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataCleanup.php:
```diff
@@ -0,0 +1,320 @@
+	 */
+	public function process_batch( array $batch ): void {
+		// This is a destructive operation, so check if we need to bail out just in case.
+		if ( ! $this->should_run() ) {
```
Maybe this check should also be added in `get_next_batch_to_process` and `get_total_pending_count` so that they both return 0 when we need to bail.

jorgeatorres replied:
Done in ad534451f8f5ed14e04a373b4abe6f0691b98182.

vedanshujain reviewed (commented):
Looks good to, and tested well too. Although I did found one issue but it might be unrelated to this PR. 

I introduced two error orders, one where I manually deleted from orders table, and another I overwrote to have new data in posts. As expected the cleanup process never completed because it got stuck into those error orders. Maybe we should write into log when that happens.

Another unrelated thing to this PR is, when I tested using CLI with the above error orders, I ran into infinite loop (let me know if you can't repro and I can debug):

![Screenshot 2024-01-29 at 8 46 00 PM](https://github.com/woocommerce/woocommerce/assets/7571618/26263ce1-7ced-425f-b890-7b1b39fbf092)

jorgeatorres committed:
Extend bail early protection to other batch processing functions

jorgeatorres committed:
Log and stop cleanup when a full batch fails

jorgeatorres committed:
Make PHPCS happy

jorgeatorres commented:
> I introduced two error orders, one where I manually deleted from orders table, and another I overwrote to have new data in posts. As expected the cleanup process never completed because it got stuck into those error orders. Maybe we should write into log when that happens.

I agree. We were already logging errors but I also added some additional logging with 2a50813b0e448d2006c25536144c948a63fa8030. Now if a full batch of orders fails to process we log that and abort the process:

<img width="1416" alt="Screenshot 2024-01-30 at 16 08 10" src="https://github.com/woocommerce/woocommerce/assets/184724/34b408bd-201b-4a78-869d-f717d259df99">

I understand there still might be orders we could process, but I didn't want to keep too much state or complicate the SQL queries involved. Aborting seemed like a good compromise as long as it's logged. wdyt?


> 
> Another unrelated thing to this PR is, when I tested using CLI with the above error orders, I ran into infinite loop (let me know if you can't repro and I can debug):

I was able to replicate. I'll log an issue about it.

- - -

@vedanshujain: could you please taking another look? As always, thanks for the review!

Konamiman commented:
LGTM and tests as advertised. @vedanshujain anything else to add?

vedanshujain reviewed (approved):
This time my build process is failing, so the loading gif is not giffing for me probably a local setup issue. Approving and merging, based on N's comment and other than the gif its working well.

luissimo commented:
So the tool cannot be used if one batch fails? There are orders that get deleted automatically after a certain period through the settings in Woocommerce -> Settings -> Accounts & Privacy. 

Now a single batch with those old order id's stops the entire process.

jorgeatorres commented:
Hi @luissimo! Thanks for the bug report. We'll take a look.

svenms commented:
Hello,

I’m trying to use the new tool added to Woocommerce to clean the old orders from post and postmeta, but unfortunately it displays an error :

Error No se pudieron limpiar los datos heredados del pedido 93102 durante el proceso por lotes. Error: 93102 no es un ID de pedido válido.

While checking in Backend Admin Orders, this order does not exist. But it does if I search it in Post table in the DB.

best regards!

jorgeatorres commented:
Hi @svenms! As you've noticed, this happens when the order no longer exists in HPOS but still exists as a post. This is a very unusual situation. We've added some code to prevent this from happening and also some enhancements to the cleanup tool so that it more gracefully handles the situation, but those fixes will be released with 8.9.0, which is planned for next month. Right now, you could just remove the post and the metadata manually for that problematic order. For that, you can use `wp post delete 93102`. Double check IDs as this a destructive operation.

svenms commented:
> Hi @svenms! As you've noticed, this happens when the order no longer exists in HPOS but still exists as a post. This is a very unusual situation. We've added some code to prevent this from happening and also some enhancements to the cleanup tool so that it more gracefully handles the situation, but those fixes will be released with 8.9.0, which is planned for next month. Right now, you could just remove the post and the metadata manually for that problematic order. For that, you can use `wp post delete 93102`. Double check IDs as this a destructive operation.

Hello @jorgeatorres.

I've just updated to 8.9.0 but the problem remains the same.

svenms commented:
> Hi @svenms!
> 
> You're correct, and apologies for that. Sadly, the fix didn't make 8.9.0 but will be in 9.0.0 for sure. The PR that introduced the fix is #46970, in case you want to give that a try.

Perfect, I've replaced the file: plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php and launch the Tool.
The Tool has successfully cleaned the orders with problems.
Just one question:
1- Launch the Tool and a process appears in Scheduled Actions (in progress).
2- After a while, the process in Scheduled Actions disappears and it does not appear in Pending. But inside Tools it still appears the Cleaning Data Loading and the Cancel Button.

While checking the Scheduled Actions I can see:
2024-05-15 19:13:37 +0000
action created
2024-05-15 19:13:37 +0000
action started via Async Request
2024-05-15 19:18:54 +0000
action marked as failed after 300 seconds. Unknown error occurred. Check server, PHP and database error logs to diagnose cause.

If I try to check in the DB one of the first orders created with legacy, the order is still un wp_post table.

jorgeatorres commented:
Hi @svenms! The cleanup tool might be heavy on some databases, and that might be the reason for the failure, but without access to logs or more details it's hard to know.
As for the disconnect between what the tool reports and the A-S job, this is due to a different issue we're working on (#44271). That said, that's just to re-try the process automatically. It doesn't explain what happens in the first place.

Do you have any logs or details you can share?

svenms commented:
Hello,

You're right. We did the process for about 8 times and it finished well.
Maybe it should be added some message on top of wordpress backend when the
process has to be terminated after 300 seconds and it has not finish.
Also the loading in Tools should be finished after 300 seconds, otherwise
people would think it is still in process.

Thank you very much and great job!

El lun, 20 may 2024 a las 11:55, Jorge A. Torres ***@***.***>)
escribió:

> Hi @svenms <https://github.com/svenms>! The cleanup tool might be heavy
> on some databases, and that might be the reason for the failure, but
> without access to logs or more details it's hard to know.
> As for the disconnect between what the tool reports and the A-S job, this
> is due to a different issue we're working on (#44271
> <https://github.com/woocommerce/woocommerce/issues/44271>). That said,
> that's just to re-try the process automatically. It doesn't explain what
> happens in the first place.
>
> Do you have any logs or details you can share?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce/pull/43293#issuecomment-2120736024>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSEQUVQOCFFCIZSESVNY3DZDIMHJAVCNFSM6AAAAABBNHQFXWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRQG4ZTMMBSGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

thinkjarvisdesignandmarketing commented:
Hi @jorgeatorres 

We have a huge store with 133,000 orders. It has so far taken over a day to remove 50,000 order records from post_meta during the Clean up order data from legacy tables action.

Can you share the SQL code that is being ran on the database so I can run it in a single action instead of async?

Or any way I can speed up the frequency of the async commands?

Konamiman commented:
Hi @thinkjarvisdesignandmarketing. The following snippet will perform the cleanup in a single run (adjust the 1000 passed to `get_next_batch_to_process` if needed):

```php
$ldc = wc_get_container()->get(Automattic\WooCommerce\Internal\DataStores\Orders\LegacyDataCleanup::class)
while(true) {
	$ids = $ldc->get_next_batch_to_process(1000);
	if(empty($ids)) break;
	$ldc->process_batch($ids);
}
```

For reference:

* [`get_next_batch_to_process`](https://github.com/woocommerce/woocommerce/blob/release/9.5/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataCleanup.php#L109)
* [`process_batch`](https://github.com/woocommerce/woocommerce/blob/release/9.5/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataCleanup.php#L120)

...which end up calling these:

* [`get_orders_for_cleanup`](https://github.com/woocommerce/woocommerce/blob/release/9.5/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php#L74)
* [`cleanup_post_data`](https://github.com/woocommerce/woocommerce/blob/release/9.5/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php#L156)

Also the SQL query that gets the orders pending cleanup (generated by [`build_sql_query_for_cleanup`](https://github.com/woocommerce/woocommerce/blob/release/9.5/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataHandler.php#L91)) is:

```sql
SELECT wp_posts.ID FROM wp_posts LEFT JOIN wp_wc_orders ON wp_posts.ID = wp_wc_orders.id 
WHERE (wp_posts.post_type IN ('shop_order', 'shop_order_refund') 
  OR (post_type = 'shop_order_placehold' 
  AND ( wp_wc_orders.id IS NULL OR EXISTS(SELECT 1 FROM wp_postmeta WHERE post_id = wp_posts.ID)) )) 
  AND wp_posts.post_status != 'auto-draft'
```
